File: rlgym_learn/standard_impl/ppo/trajectory_step.py
from dataclasses import dataclass
from typing import Generic, Optional

from rlgym.api import ActionType, AgentID, ObsType, RewardType
from torch import Tensor


@dataclass
class TrajectoryStep(Generic[ObsType, ActionType, RewardType]):
    __slots__ = ("obs", "action", "log_prob", "reward", "value_pred")
    obs: ObsType
    action: ActionType
    log_prob: Tensor
    reward: RewardType
    value_pred: Optional[Tensor]

File: src/standard_impl/ppo/trajectory_step.rs
use pyo3::prelude::*;
use pyo3::PyObject;

#[allow(dead_code)]
#[derive(FromPyObject)]
pub struct TrajectoryStep {
    pub obs: PyObject,
    pub action: PyObject,
    pub log_prob: PyObject,
    pub reward: PyObject,
    pub value_pred: Option<PyObject>,
}
